### ISSUE TITLE
Idea: externalize all side-effects just like user interaction is

### ISSUE THREAD

**Issue by staltz** [staltz/cycle#116]
We could generalize interactions with external systems, and make the user interaction to be just one of those interactions. For instance, human-computer interaction happening simultaneously with server-computer interaction.

Proposed API:
```js
function computer(interactions) {
  var userClicks = interactions.get('dom://.target/click');
  var responses = interactions.get('mongodb://foods/3/best-before');
  // ...
  return {
    vtree$, // as usual
    request$ // for mongodb
  };
}

Cycle.apply(computer, {
  dom: '#app',
  mongodb: someMongoDriverHereWithConnectionConfig
});
```

**Comment by ivan-kleshnin** [comment]
:+1: the world doesn't end on DOM. Even on client Document title is not DOM. URL is not DOM. Audio API is not DOM. We can continue this forever. Do you think Cycle could enter backend area? I do!

**Comment by staltz** [comment]
> Do you think Cycle could enter backend area?

Yes! But Cycle right now is 80% DOM specific. For the backend, I think Streams/inject would be enough.

**Comment by schrepfler** [comment]
We have in mind something similar where the external data source will be something a-la-falcor

**Comment by erykpiast** [comment]
Seems interesting at least. I guess the best in this kind is to generalize whole concept of side effect and do it like:

```javascript
Cycle.apply(computer, {
  dom: cycleDomAdapter( '#app'),
  mongodb: cycleMongoAdapter(...)
});
```
Adapter has to implement universal interface (I'm not sure if  `{ get(param1, param2) }` needs to be obligatory signature) and that's all. Cycle itself could be 100% environment-agnostic. Just an idea, really don't know how it could work in practice ;)

**Comment by staltz** [comment]
These are good ideas, I liked your `cycleDomAdapter('#app')`. I was just thinking about this some minutes ago, and seems like it should be the way to go. And I prefer `get(uri)` instead of `get(param1, param2)`, where `uri` is of the format `context://some/path/to/some/observable/resource`. For instance, because we registered `dom: cycleDomAdapter('#app')`, we would use it as `interactions.get('dom://.target/click')`.

Because we still have `computer()` and `user()` functions (the latter is implicit inside `applyToDOM`), it should be that `cycleDOMAdapter()` returns simply a function, namely, it will return precisely the `user()` function. I still need to figure out exactly what contract should the functions returned by these adapters look like.

**Comment by staltz** [comment]
Proposed name for user() and others functions:  "external functions". Any function that the computer calls would be an "internal" function to your application, but others are "external" to the application. Might live remotely (e.g. user, database), or might live in the same JavaScript context (as an adapter to Audio APIs or other Web APIs not related to the UI DOM).

External functions are normally assembled from an object of configurations, for example:
```js
let user = makeDOMUser('#app', 'vtree$');
```

`makeDOMUser(container, vtreeObservableName)` seems to work well, where `vtreeObservableName` is optional, if not given we assume `vtree$` if the `computer`'s output is an object, else if the computer's output is an Observable, then we assume it is the vtree Observable.

Another example:
```
let mongo = makeMongoDriver({
  host: 'foobar.io', 
  port: '3000', 
  user: 'foo', 
  password: 'secret', 
  requestObservable: 'request$'
});
```

And usage:
```js
Cycle.run(computer, {
  dom: user,
  mongodb: mongo,
});
```

`makeFoobar()` should be a factory for a function. The function assumes `computerOutput :: Object` as an input, and output a queryable (`get`) object of Observable. More details about this later

**Comment by erykpiast** [comment]
:+1:  Why don't want you to make it whole based on single key? I mean, with code like this
```javascript
Cycle.apply(computer, {
  dom: cycleDomAdapter( '#app'),
  mongodb: cycleMongoAdapter(...)
});
```
observable name for DOM adapter could be `dom$` and for Mongo `mongodb$`, automatically. I know, we loose semantics (vtree isn't a DOM actually, mongodb is the name of database, not the thing we want to operate on - requests to DB), but thanks to this approach we can reduce boilerplate and, what maybe is a separate issue, avoid accidental defaults. `vtree$` as a default value would look like very weak argument in the future, when Cycle becomes industry-standard framework for everything ;) "You know, back then it supposed to be front-end only framework, was based on VirtualDOM, because native one was painfully slow, it's our legacy..." etc. It's always embarrassing to have to say something like this :)

**Comment by staltz** [comment]
Hmm, I'll consider that. thanks

**Comment by staltz** [comment]
> observable name for DOM adapter could be dom$ and for Mongo mongodb$, automatically. I know, we loose semantics (vtree isn't a DOM actually, mongodb is the name of database, not the thing we want to operate on - requests to DB), but thanks to this approach we can reduce boilerplate and, what maybe is a separate issue, avoid accidental defaults. vtree$ as a default value would look like very weak argument in the future, when Cycle becomes industry-standard framework for everything.

@erykpiast I found a tricky corner case: what about vtree$ returned by a custom element? Custom element definitions don't need to call `Cycle.applyToDOM` because it's implicit. Any suggestions? Keep vtree$ convention or change that somehow? Custom elements should return an object with vtree$ and other observables which will be recognized as custom events.

**Comment by erykpiast** [comment]
Hmm, it depends, what concept custom elements represent.

If we consider them as a simply as containers for some repeated DOM pieces, they should be part of DOM adapter and there is no need to care about them when designing adapters API. `vtree$` could be default for custom elements and it wouldn't be inconsistent in any way with `.apply` API. I suppose, that within each adapter there will be a lot of concepts and conventions different from ones in Cycle core. You just can't create single set of rules (on this level of specificity like naming or default values) convenient for all tasks you can imagine.

On the other hand, if custom elements are main building blocks of Cycle app, we would change their API too. Let's say you want to create component like [this one](https://www.polymer-project.org/0.5/docs/elements/core-ajax.html) using Cycle. It doesn't render anything, just takes some input (through attributes) and returns some output (through events). In that case you wouldn't like to create empty `vtree$` just to satisfy API requirements.

The second concept is still a little bit vague for me, I have to think about it more and make it clear. I'm not sure if those render-less components should even exist in Cycle world, we have functions in JavaScript, haven't we?

Anyway, what I want to do, is just to ask a simple question: should we consider custom elements as a DOM wrappers, or main building blocks in Cycle app? If we'll answer for that, we know the rest :)

**Comment by staltz** [comment]
Custom elements as DOM wrappers, definitely I'm not into custom elements as abstract building blocks like many Polymer elements are.

**Comment by erykpiast** [comment]
So, in my opinion default observable name can be different for custom elements and adapters.

**Comment by staltz** [comment]
Alright

**Comment by staltz** [comment]
Started working on this :+1:

**Comment by ggoodman** [comment]
What value does the use of URIs add?

Why force URIs on adapaters. Couldn't you delegate stream creation to adapters in a more flexible way like:

```javascript
function app(ext) {
  return {
    dom: ext.get$('dom', {selector: '.myinput', event: 'change'})
      .map(ev => ev.target.value)
      .startWith('')
      .map(name =>
        h('div', [
          h('label', 'Name:'),
          h('input.myinput', {attributes: {type: 'text'}}),
          h('hr'),
          h('h1', `Hello ${name}`)
        ]);
      })
  };
}

Cycle.run(app, {
  dom: Cycle.makeDOMAdapter('.js-container')
});
```

The DOM adapter may require parameters that serialize poorly to the URI scheme and just introduces the need to mentally serialize and then deserialize via code.

**Comment by staltz** [comment]
@ggoodman Thanks, I'll consider your critique

**Comment by staltz** [comment]
From gitter:

```js
function app(ext) {
  var input$ = ext.get('foo', 'myChannel');
  // ...
  return {
    foo: someObservable
  };
}

Cycle.run(app, {
  foo: wsAdapter
});
```

Implementation of an adapter function (or function factory), e.g. wsAdapter, would rougly look like your app function. It takes the `foo` (returned by the app function) as input and should produce a queryable object wrapping some Observables.

For the input, you don't need to do `appObservables.foo`, because Cycle.run() handles that for you. It will be just `function wsAdapter(input)` and Cycle.run() will give the app's `foo` output as `input` for wsAdapter.

For the output of wsAdapter, I'm considering 3 alternatives (all 3 would be available for adapter developers): 

1.  Return an object with a `get` method
2.  Return an object with Observables and their names
3.  Return just one Observable

To use case (1) in myApp: `ext.get('foo', param1, param2, etc)`
To use case (2) in myApp: `ext.get('foo', 'nameOfObservableInThatObject')`
To use case (3) in myApp: `ext.get('foo')`

**Comment by alexmingoia** [comment]
I would lean towards simplicity (3) without understanding the use case for (1) and (2). Why is it important to be able to return anything but one observable for the adapter output?

**Comment by cgeorg** [comment]
@alexmingoia How would that work for something like the domAdapter, which has infinitely many events it could be emitting?

**Comment by staltz** [comment]
@alexmingoia I'm thinking all 3 cases would be supported, not *one* of them. And as @cgeorg said, dom adapter has many user interaction Observables.

**Comment by staltz** [comment]
Naming problem: I'm not entirely satisfied with the name "adapter" for these things. Neither am I satisfied with "driver". "Adapter" is technically correct, because we adapting the DOM's API (or Mongo's API, or whatnot) to conform to `fn(Observable<A> input): Observable<B>` contract. However for someone trying to understand Cycle, these adapters are just "members of an interaction with the app function". I want to keep the mental model of "computer() function interacting with the user() function, like a conversation". The word I found that most accurately represents this is *[Interactant](http://dictionary.reference.com/browse/interactant)* but I'm a bit concerned this is such a different word to use.

Any opinions to help me get unstuck?

**Comment by trxcllnt** [comment]
@staltz *Wario voice* the IO Monad strikes again. Mwahaha!

**Comment by schrepfler** [comment]
How about bridge although there's a design pattern behind that term?
'''
Cycle.run(app, {dom: makeDOMBridge('.container')});
'''

**Comment by staltz** [comment]
@trxcllnt care to explain the connections between this and monads? I had a hint a clue there was, but not enough courage and knowledge to claim that.

**Comment by tel** [comment]
There's something a little similar between this and the GoF "interpreter" pattern. You can see this arising in "Free" monads which are a good type for expressing reified instruction for an interpreter. I'm not personally confident that the connection is so strong, though...

**Comment by staltz** [comment]
@tel I can see the similarity between e.g. the DOM adapter here and the interpreter pattern with regards to *writes* done by the DOM adapter. After all the DOM adapter interprets VTree data structure and then proceeds to evaluate that to the DOM. But the "reading from the DOM" part isn't normally solved by an interpreter.

**Comment by tel** [comment]
You can write interpreters in that pattern by more or less embedding parallel threads into your interpreted language. Not much different than setting functions as callbacks on event handlers.

**Comment by erykpiast** [comment]
Well, while `Interactant` describes purpose of "those things" very well, I guess it's really to uncommon word in average programmer's dictionary (no data behind it, only good feeling!) and it will seem weird for newcomers. Not "new", not "unknown" but just weird. It's my opinion.

I'm not saying we should keep "adapter" as the official name. I guess I've used this term first, but it was really unthought, the first word that came to my mind. So, I always feel bad when my criticism is not constructive, so I oblige myself to propose something else :D But lets deconstruct the problem first.

We need a name for the thing that:

 * is a source of interactions with a computer (through network, input devices, whatever)
 * is a sink to which we pour data (their representation is visible in case of DOM, but for WS or XHR it's received by server)

When I think about it like this, only one term comes to my mind: IO... something. Driver? ~~Adapter~~? Device? Bridge?

And regarding function names, for factory I always opt on `createSomething` for factories and single nouns for constructors. Anyway, I'm not sure we should force "those things" names, by design they are created by community, so I guess everybody will use his favourite pattern. Personally, I prefer factories.

**Comment by jessehattabaugh** [comment]
I vote Adapter. It's commonly used in frameworks to refer to plugins that help the framework talk to things it wasn't originally designed to work with, and it seems like this is how Adapter writers will need to think of them. 

I only have a loose grasp of what these are for, but if I'm understanding what you mean when you say they are "members of an interaction with the app function" then how about "Agent", or "Broker". These are the least overloaded words I could think of.

**Comment by staltz** [comment]
Thanks for the suggestions above, guys.
The reason why I suggested "interactant" (and I'm not necessarily still advocating it yet) is because I find it important that the app function is symmetric to "adapters". The app function itself is an adapter. The point being: both app function and "adapters" should be conceptually named the same thing. Calling the `app` an adapter isn't intuitive. `app` as a Bridge might work. `app` as an Agent might work. And maybe there are similarities with [Arrows](https://en.wikipedia.org/wiki/Arrow_(computer_science)).

**Comment by tel** [comment]
Here's a term which is well-known in another field but I never see in CS: "Plant".

https://en.wikipedia.org/wiki/Plant_%28control_theory%29

The article is a little weak, but the idea is that a plant is the component in a control system which bounds the exterior world. The control system reads information from sensors and from the plant itself and then sends the "proper" control signal to the plant itself to actuate in the exterior world.

I think this is a direct analogue to what is being discussed.

**Comment by danyx23** [comment]
Tough one. I think Adapter is not too bad. I personally think Interactant is a bit too foreign, but that's just me.

As for other ideas: Some good names for design patterns came from performing arts, maybe this could help here. This one is probably a bit far out, but just to give new input for brainstorming: Puppet. You make it do something but it also reacts back, it is a standin for the interaction between the puppet performer and the audience - where the performer would be the app and the audience the user or the server or whatever.

**Comment by schrepfler** [comment]
Heh, small side joke, just like GoF had Observer and Rx introduced Observable, maybe yours can be called Bridgeable/Adaptable :)

**Comment by ArnoBuschmann** [comment]
For the sake of creativity -> brainstorming -> http://en.wikipedia.org/wiki/Brainstorming

- defer judgement
- focus on quantity

adapter, interactant, driver, bridge, hub, communicator, plug, connector, collaborator, merger, joiner, combiner, joint, junction, link, caster, broadcaster, service, middleware (it's not Express? oh wait, that's a judgement... ;)

- combine ideas

communication plug, adaptive merger, streaming link, data-jockey, stream-juggler...

This is just a suggestion for the readjustment process, not a serious try to suggest the "right name" in the first place.

**Comment by staltz** [comment]
:+1: @ArnoBuschmann

**Comment by staltz** [comment]
> adapter, interactant, driver, bridge, hub, communicator, plug, connector, collaborator, merger, joiner, combiner, joint, junction, link, caster, broadcaster, service, middleware, ...

thing, person, talker, member, participant, speaker, orator, sparring assistant, communicator, individual, node, peer, ...

**Comment by alexmingoia** [comment]
Call them "cycles" or "cyclers" hehe

**Comment by jessehattabaugh** [comment]
Oooo, bicycle metaphors! Gears, Sprockets, Chains, Cables, Pedals

**Comment by ArnoBuschmann** [comment]
cyclers -> makes me smile,
pedals -> nice,
wheels?

Sounds a bit like should a name being constrictively descriptive or ambiguous which could be good as the concept behind has to be explained anyway?

**Comment by alexmingoia** [comment]
My final suggestions in order of preference:

1. adapter
2. interaction
3. channel

**Comment by staltz** [comment]
Gear!
Has 4 letters (I like Communicator but it's too damn long), has been used as a synonym for plugin in other systems, matches the idea and diagram of a cycle, and most of all sounds cool and familiar.

**Comment by jessehattabaugh** [comment]
Why can't they be Streams or Channels? Are those reserved words?

**Comment by staltz** [comment]
They are functions from Observable to Observable. Another word for Observable is Stream, so calling the function a Stream would be really bad conflation.

**Comment by ggoodman** [comment]
Another word that comes to mind is Bus.

**Comment by cgeorg** [comment]
I think Bridge, Interpreter and Translator all do a good job of supporting the idea of a conversation between 2 or more functions.  Adapter brings to mind a transformation (of Observable to Observable) so that is a solid choice if that is what you want.  Most other things that come to mind lean more towards producer/consumer, and not the combination of both.

Gear is... short.  Cog is shorter.  I don't think either of them clearly communicate the idea of the conversation.  Smaller components of a larger system perhaps.  "DomGear" though?  No thanks.

I don't think these words will be typed too often - there is only one Cycle.run() - so I would favor clarity over brevity.  Thinking of real world names - "DomAdapter", "WebSocketAdapter", "AudioAdapter" - I prefer Adapter to the other suggestions.  Interactant is also pretty good, just not familiar.  From some above suggestions, Link and Connector would also be ok.

**Comment by staltz** [comment]
> Most other things that come to mind lean more towards producer/consumer, and not the combination of both.

Yes, this is an important aspect.

> "DomGear" though? No thanks.

Sadly, I was already happy about "DOMGear". I find it important feedback that other people don't like it.

**Comment by jessehattabaugh** [comment]
Hate to admit it since I brought it up, but I'm not that crazy about "Gears" either. It makes me think of that XBox game "Gears of War" where all the soldiers we called [Gears](http://gearsofwar.wikia.com/wiki/Gear) because they were supposed to be part of a war machine or something. But more importantly I don't think it communicates what they *are*, or what their role in the system is. 

My impression is that these are interfaces where events can flow in, and vdom trees can flow out. If that's the case I really think that **Channel** is the most apt word for them. Like channels on a radio dial where messages pass through. Conversations are carried out on these frequencies, and you can listen to one or more of them at a time (if you have two radios I suppose :-)

I know that "Channel" is pretty overused (especially in Web Socket land), so what about some more cycling metaphors! Bike **Paths** are kinda like channels that carry bikes. These are called Cycle **Lanes** in some countries. Roads can have multiple Lanes that traffic flows through. In order to get to certain places you need to get in the right lane. If you want certain traffic to get to a certain place you make sure that the Lane goes there. 

DomLane, SocketLane, ChatLane, IntentLane, ModelLane, ViewLane, NativeLane, VoiceCommandLane, ExpressLane, RestLane, PersistenceLane, DatabaseLane, IndexDBLane, etc

**Comment by staltz** [comment]
Ok, I'm starting to put down "Gears". And I guess "Cog" goes too.

Channel is too overused. Path too. Lane isn't, but I'm not sure it sounds natural to have a conversation with a Lane, or to interact with a Lane.

Best candidates so far are:
- Adapter
- Interactant

**Comment by staltz** [comment]
New ones brought to the table inspired by mutualism and symbiosis:

- Symbiont
- Mutualist

Of which, "Symbiont" I like most.

**Comment by ArnoBuschmann** [comment]
- Satellite

I'd go with adapter and try to stay focused on introcuing cycle as the main new concept. Everyone knows what an adapter is and that it is useful for connecting something. Adapter is self-explanatory while symbiont/mutualist/a-lot-of-other-names have to be explained.

**Comment by staltz** [comment]
Thanks for the feedback

**Comment by staltz** [comment]
Ok, I think I decided a name and I feel quite confident about this one: "driver". It was the first name I used, but I dropped when I didn't want to make it sound like an OS Driver. However, reading how Haskell 1.0 was very similar to Cycle, they had a cyclic interaction between `app` and `os`, which makes a lot of sense. In Cycle, one can consider the "operating system" to be the proxy between the app and the user. The DOM is in many ways the operating system of the web. And Cycle isn't just for the web. So it also makes sense to use the word "driver" as adapters/interfaces to hardware or accessories which enable interaction with the outside world, whether that be a user or another computer (a server).

**Comment by serapath** [comment]
I like the most: `Streams`, `Channels`, `Producer`, `Consumer`, `Satelite`, `Interactant`.
When I was playing around, I introduced names like `User$`, `System$` to differentiate between an event stream coming in from a user compared to a backend or other part of "the system".

Although, I have to admit, I still have trouble wrapping my head around cyclejs, because even though i like the presentation from @staltz about this computer() and human() and it all makes a lot of sense to me.
1. I have trouble fitting the DOM as an "ambassador" or something along those lines into the picture and the virtual dom. I always feel, this is something that should be part of computer()...
2. I don't like the global functions or method from the "Cycle.xyz" object, they give me the impression, that the whole app is kind of "managed" by the "cycle" thing.

**Question:** How would I create a micro micro computer() function (that encapsulates something like a component) and make a second one and then "wire them up" to each other. Both potentially receiving input from `user(s)` or `system(s)`.... which, I assume, are both somehow "Interactants", right?

Please forgive me if I might be completely off here :-)

**Comment by jessehattabaugh** [comment]
Driver +1

**Comment by staltz** [comment]
Now in master. Soon released as v0.23.

**Comment by RangerMauve** [comment]
@serapath In terms of your question, it should be easy enough to just create a function that takes in whatever interactions or system resources it relies on and outputs whatever it is that the rest of the app requires. For instance if you wanted to listen on UI ques or on server sent events  to trigger a refresh, you could just make a functuon that takes the observables for the interactions and the server sent events, and returns an observable for triggering the refresh.
